### PR TITLE
[SystemC] Introduce special types for ports and signals

### PR DESCRIPTION
**Pull request by maerhart**
Remove the current port direction attribute and enum and use types to represent the port direction. Reuse the PortInfo struct and enum of the HW dialect to represent the port direction in CPP code when MLIR types aren't that well suited. The rational is that these constructs are the same across multiple dialects and should be part of the ModuleLike interface to be reused by all module ops implementing the HWModuleLike interface. Providing getters and builders that accept these port direction/port info struct makes conversion between module ops that implement HWModuleLike very convenient. This PR also implements that interface for SystemC modules.
Also switch over to use the module_like and function_like parser and printer utilities and add support for module visibility and argument attributes.

### CODE REVIEW

File: include/circt/Dialect/SystemC/SystemCExpressions.td [hunk 1]
```diff
@@ -10,3 +10,27 @@
 //
 //===----------------------------------------------------------------------===//
 
+def SignalReadOp : SystemCOp<"signal.read", [
+    NoSideEffect,
+    TypesMatchWith<"result has to be the base type of the input",
+                   "input", "result", "systemc::getSignalBaseType($_self)">]> {
```
Comment (seldridge) [inline include/circt/Dialect/SystemC/SystemCExpressions.td:16]: Would using `InferTypeOpInterface` simplify this?

File: include/circt/Dialect/SystemC/SystemCTypesImpl.td [hunk 2]
```diff
@@ -21,6 +21,60 @@ def FuncHandleTypeImpl : SystemCTypeDef<"FuncHandle"> {
...
+}
+
+// Represents the SystemC sc_out<T> type.
+// Declares the systemc::OutputType in C++.
+def OutputTypeImpl : SignalTypesImplBase<"Output"> {
+  let summary = "Represents a SystemC sc_out<T> value.";
+  let description = [{
+    Represents the specialized SystemC port class sc_out<T> as described in
+    IEEE 1666-2011 §6.12.
+  }];
```
Comment (seldridge) [inline include/circt/Dialect/SystemC/SystemCTypesImpl.td:67]: Thank for your painstakingly adding all the section information to the ODS. 🙏

File: lib/Dialect/SystemC/SystemCOps.cpp [hunk 3]
```diff
@@ -33,11 +35,29 @@ static void printImplicitSSAName(OpAsmPrinter &p, Operation *op,
 // SCModuleOp
 //===----------------------------------------------------------------------===//
 
-void SCModuleOp::getPortsOfDirection(PortDirection direction,
-                                     SmallVector<Value> &outputs) {
+static hw::PortDirection getDirection(Type type) {
+  return TypeSwitch<Type, hw::PortDirection>(type)
+      .Case<InOutType>([](auto ty) { return hw::PortDirection::INOUT; })
+      .Case<InputType>([](auto ty) { return hw::PortDirection::INPUT; })
+      .Case<OutputType>([](auto ty) { return hw::PortDirection::OUTPUT; });
```
Comment (seldridge) [inline lib/Dialect/SystemC/SystemCOps.cpp:42]: Should this defensively code a default with an assert?

**Review by seldridge — APPROVED** [review]
Nice. This looks way cleaner now.

File: lib/Dialect/SystemC/SystemCOps.cpp (hunk 3 above)
Comment (maerhart) [reply lib/Dialect/SystemC/SystemCOps.cpp:42]: `TypeSwitch` already has an assert when no case matches. The enclosing function is part of the stack trace and thus easy to find.

Comment (dtzSiFive) [reply lib/Dialect/SystemC/SystemCOps.cpp:42]: Oh, TIL!  FWIW it appears this is only the behavior when accessing the return value-- using the void specialization (`TypeSwitch<Type>`) or even using `TypeSwitch<Type,int>` won't assert on missed cases if you ignore the return value.
This seems reasonable, certainly for the void case, but good to know.  Anyway, thanks!

**Pull request merged**